[Bug] panic when using audit webhook

### Bug Description

Got panic when using flipt with audit webhook is configured. This cause server unavailable and audit function does not work properly

### Version Info

v1.46.0

### Search

- [X] I searched for other open and closed issues before opening this

### Steps to Reproduce

Config follow https://github.com/flipt-io/flipt/blob/v1.46.0/examples/audit/webhook/docker-compose.yml
When create flag via UI, got below panic and process is not reachable anymore

```
webhook-flipt-1    | panic: invalid logger type passed, must be Logger or LeveledLogger, was *zap.Logger
webhook-flipt-1    | 
webhook-flipt-1    | goroutine 135 [running]:
webhook-flipt-1    | github.com/hashicorp/go-retryablehttp.(*Client).logger.func1()
webhook-flipt-1    | 	github.com/hashicorp/go-retryablehttp@v0.7.7/client.go:463 +0xcd
webhook-flipt-1    | sync.(*Once).doSlow(0x487c40?, 0xc0000f33f0?)
webhook-flipt-1    | 	sync/once.go:74 +0xc2
webhook-flipt-1    | sync.(*Once).Do(...)
webhook-flipt-1    | 	sync/once.go:65
webhook-flipt-1    | github.com/hashicorp/go-retryablehttp.(*Client).logger(0xc0000f3380)
webhook-flipt-1    | 	github.com/hashicorp/go-retryablehttp@v0.7.7/client.go:453 +0x45
webhook-flipt-1    | github.com/hashicorp/go-retryablehttp.(*Client).Do(0xc0000f3380, 0xc000acea08)
webhook-flipt-1    | 	github.com/hashicorp/go-retryablehttp@v0.7.7/client.go:656 +0x9b
webhook-flipt-1    | go.flipt.io/flipt/internal/server/audit/webhook.(*webhookClient).SendAudit(0xc0008c4ff0, {0x3b16400, 0xc000372e70}, {{0x304c485, 0x3}, {0x304cd89, 0x4}, {0x305204e, 0x7}, {0xc0008b6730}, ...})
webhook-flipt-1    | 	go.flipt.io/flipt/internal/server/audit/webhook/client.go:72 +0x32f
webhook-flipt-1    | go.flipt.io/flipt/internal/server/audit/webhook.(*Sink).SendAudits(0xc000b32fd8, {0x3b16400, 0xc000372e70}, {0xc000372f50?, 0x2?, 0x2?})
webhook-flipt-1    | 	go.flipt.io/flipt/internal/server/audit/webhook/webhook.go:31 +0x119
webhook-flipt-1    | go.flipt.io/flipt/internal/server/audit.(*SinkSpanExporter).SendAudits(0xc0008d4d40, {0x3b16400, 0xc000372e70}, {0xc000372f50, 0x1, 0x1})
webhook-flipt-1    | 	go.flipt.io/flipt/internal/server/audit/audit.go:85 +0x2e9
webhook-flipt-1    | go.flipt.io/flipt/internal/server/audit.(*SinkSpanExporter).ExportSpans(0xc0008d4d40, {0x3b16400, 0xc000372e70}, {0xc0008d4d60?, 0xc000b2bef2?, 0xc0009c2540?})
webhook-flipt-1    | 	go.flipt.io/flipt/internal/server/audit/audit.go:60 +0xd9
webhook-flipt-1    | go.opentelemetry.io/otel/sdk/trace.(*batchSpanProcessor).exportSpans(0xc0006f61e0, {0x3b16390, 0xc000858140})
webhook-flipt-1    | 	go.opentelemetry.io/otel/sdk@v1.27.0/trace/batch_span_processor.go:277 +0x238
webhook-flipt-1    | go.opentelemetry.io/otel/sdk/trace.(*batchSpanProcessor).processQueue(0xc0006f61e0)
webhook-flipt-1    | 	go.opentelemetry.io/otel/sdk@v1.27.0/trace/batch_span_processor.go:321 +0x308
webhook-flipt-1    | go.opentelemetry.io/otel/sdk/trace.NewBatchSpanProcessor.func1()
webhook-flipt-1    | 	go.opentelemetry.io/otel/sdk@v1.27.0/trace/batch_span_processor.go:117 +0x54
webhook-flipt-1    | created by go.opentelemetry.io/otel/sdk/trace.NewBatchSpanProcessor in goroutine 1
webhook-flipt-1    | 	go.opentelemetry.io/otel/sdk@v1.27.0/trace/batch_span_processor.go:115 +0x2e5
```

### Expected Behavior

Webhook should be fired and server still be healthy

### Additional Context

Docker

This appears to have resurfaced in `1.48.0`
"docker.flipt.io/flipt/flipt:latest"

```
panic: invalid logger type passed, must be Logger or LeveledLogger, was *zap.Logger
goroutine 74 [running]:
github.com/hashicorp/go-retryablehttp.(*Client).logger.func1()
	github.com/hashicorp/go-retryablehttp@v0.7.7/client.go:463 +0xcd
sync.(*Once).doSlow(0x487c40?, 0xc0006f6b70?)
	sync/once.go:74 +0xc2
sync.(*Once).Do(...)
	sync/once.go:65
github.com/hashicorp/go-retryablehttp.(*Client).logger(0xc0006f6b00)
	github.com/hashicorp/go-retryablehttp@v0.7.7/client.go:453 +0x45
github.com/hashicorp/go-retryablehttp.(*Client).Do(0xc0006f6b00, 0xc000926738)
	github.com/hashicorp/go-retryablehttp@v0.7.7/client.go:656 +0x9b
go.flipt.io/flipt/internal/server/audit/template.(*webhookTemplate).Execute(0xc000a77410, {0x3ed7c60, 0xc000310150}, {{0x33aae34, 0x3}, {0x33ab70e, 0x4}, {0x33b1114, 0x7}, {0xc0009caaa0}, ...})
	go.flipt.io/flipt/internal/server/audit/template/executer.go:96 +0x4b2
go.flipt.io/flipt/internal/server/audit/template.(*Sink).SendAudits(0xc000d124c0, {0x3ed7c60, 0xc000310150}, {0xc000310380?, 0x2?, 0x2?})
	go.flipt.io/flipt/internal/server/audit/template/template.go:46 +0x155
go.flipt.io/flipt/internal/server/audit.(*SinkSpanExporter).SendAudits(0xc000d12500, {0x3ed7c60, 0xc000310150}, {0xc000310380, 0x1, 0x1})
	go.flipt.io/flipt/internal/server/audit/audit.go:85 +0x2e9
go.flipt.io/flipt/internal/server/audit.(*SinkSpanExporter).ExportSpans(0xc000d12500, {0x3ed7c60, 0xc000310150}, {0xc000d12520?, 0xc0000a8ef2?, 0xc000bb21c0?})
	go.flipt.io/flipt/internal/server/audit/audit.go:60 +0xd9
go.opentelemetry.io/otel/sdk/trace.(*batchSpanProcessor).exportSpans(0xc000be0140, {0x3ed7bf0, 0xc000b92050})
	go.opentelemetry.io/otel/sdk@v1.28.0/trace/batch_span_processor.go:277 +0x238
go.opentelemetry.io/otel/sdk/trace.(*batchSpanProcessor).processQueue(0xc000be0140)
	go.opentelemetry.io/otel/sdk@v1.28.0/trace/batch_span_processor.go:321 +0x308
go.opentelemetry.io/otel/sdk/trace.NewBatchSpanProcessor.func1()
	go.opentelemetry.io/otel/sdk@v1.28.0/trace/batch_span_processor.go:117 +0x54
created by go.opentelemetry.io/otel/sdk/trace.NewBatchSpanProcessor in goroutine 1
	go.opentelemetry.io/otel/sdk@v1.28.0/trace/batch_span_processor.go:115 +0x2e5
```

--- a/internal/cmd/grpc.go
+++ b/internal/cmd/grpc.go
@@ -16,7 +16,6 @@ import (
 	"go.opentelemetry.io/contrib/propagators/autoprop"
 
 	sq "github.com/Masterminds/squirrel"
-	"github.com/hashicorp/go-retryablehttp"
 	"go.flipt.io/flipt/internal/cache"
 	"go.flipt.io/flipt/internal/cache/memory"
 	"go.flipt.io/flipt/internal/cache/redis"
@@ -383,10 +382,9 @@ func NewGRPCServer(
 	}
 
 	if cfg.Audit.Sinks.Webhook.Enabled {
-		httpClient := retryablehttp.NewClient()
-
+		maxBackoffDuration := 15 * time.Second
 		if cfg.Audit.Sinks.Webhook.MaxBackoffDuration > 0 {
-			httpClient.RetryWaitMax = cfg.Audit.Sinks.Webhook.MaxBackoffDuration
+			maxBackoffDuration = cfg.Audit.Sinks.Webhook.MaxBackoffDuration
 		}
 
 		var webhookSink audit.Sink
@@ -394,13 +392,8 @@ func NewGRPCServer(
 		// Enable basic webhook sink if URL is non-empty, otherwise enable template sink if the length of templates is greater
 		// than 0 for the webhook.
 		if cfg.Audit.Sinks.Webhook.URL != "" {
-			webhookSink = webhook.NewSink(logger, webhook.NewWebhookClient(logger, cfg.Audit.Sinks.Webhook.URL, cfg.Audit.Sinks.Webhook.SigningSecret, httpClient))
+			webhookSink = webhook.NewSink(logger, webhook.NewWebhookClient(logger, cfg.Audit.Sinks.Webhook.URL, cfg.Audit.Sinks.Webhook.SigningSecret, maxBackoffDuration))
 		} else if len(cfg.Audit.Sinks.Webhook.Templates) > 0 {
-			maxBackoffDuration := 15 * time.Second
-			if cfg.Audit.Sinks.Webhook.MaxBackoffDuration > 0 {
-				maxBackoffDuration = cfg.Audit.Sinks.Webhook.MaxBackoffDuration
-			}
-
 			webhookSink, err = template.NewSink(logger, cfg.Audit.Sinks.Webhook.Templates, maxBackoffDuration)
 			if err != nil {
 				return nil, err

--- a/internal/server/audit/template/executer.go
+++ b/internal/server/audit/template/executer.go
@@ -51,7 +51,7 @@ func NewWebhookTemplate(logger *zap.Logger, url, body string, headers map[string
 	}
 
 	httpClient := retryablehttp.NewClient()
-	httpClient.Logger = logger
+	httpClient.Logger = NewLeveledLogger(logger)
 	httpClient.RetryWaitMax = maxBackoffDuration
 
 	return &webhookTemplate{

--- a/internal/server/audit/template/executer_test.go
+++ b/internal/server/audit/template/executer_test.go
@@ -7,6 +7,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/hashicorp/go-retryablehttp"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
@@ -26,6 +27,9 @@ func TestConstructorWebhookTemplate(t *testing.T) {
 
 	assert.Equal(t, "https://flipt-webhook.io/webhook", template.url)
 	assert.Nil(t, template.headers)
+	assert.Equal(t, 15*time.Second, template.httpClient.RetryWaitMax)
+	_, ok = template.httpClient.Logger.(retryablehttp.LeveledLogger)
+	assert.True(t, ok)
 }
 
 func TestExecuter_JSON_Failure(t *testing.T) {

--- a/internal/server/audit/template/executer_test.go
+++ b/internal/server/audit/template/executer_test.go
@@ -5,16 +5,15 @@ import (
 	"net/http"
 	"net/http/httptest"
 	"testing"
-	"text/template"
 	"time"
 
-	"github.com/hashicorp/go-retryablehttp"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 
 	"go.flipt.io/flipt/internal/server/audit"
 	"go.flipt.io/flipt/rpc/flipt"
 	"go.uber.org/zap"
+	"go.uber.org/zap/zaptest"
 )
 
 func TestConstructorWebhookTemplate(t *testing.T) {
@@ -30,14 +29,11 @@ func TestConstructorWebhookTemplate(t *testing.T) {
 }
 
 func TestExecuter_JSON_Failure(t *testing.T) {
-	tmpl, err := template.New("").Parse(`this is invalid JSON {{ .Type }}, {{ .Action }}`)
+	tmpl := `this is invalid JSON {{ .Type }}, {{ .Action }}`
+
+	template, err := NewWebhookTemplate(zaptest.NewLogger(t), "https://flipt-webhook.io/webhook", tmpl, nil, time.Second)
+
 	require.NoError(t, err)
-
-	template := &webhookTemplate{
-		url:          "https://flipt-webhook.io/webhook",
-		bodyTemplate: tmpl,
-	}
-
 	err = template.Execute(context.TODO(), audit.Event{
 		Type:   string(flipt.SubjectFlag),
 		Action: string(flipt.ActionCreate),
@@ -47,12 +43,10 @@ func TestExecuter_JSON_Failure(t *testing.T) {
 }
 
 func TestExecuter_Execute(t *testing.T) {
-	tmpl, err := template.New("").Parse(`{
+	tmpl := `{
 		"type": "{{ .Type }}",
 		"action": "{{ .Action }}"
-	}`)
-
-	require.NoError(t, err)
+	}`
 
 	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		w.WriteHeader(200)
@@ -60,11 +54,8 @@ func TestExecuter_Execute(t *testing.T) {
 
 	t.Cleanup(ts.Close)
 
-	template := &webhookTemplate{
-		url:          ts.URL,
-		bodyTemplate: tmpl,
-		httpClient:   retryablehttp.NewClient(),
-	}
+	template, err := NewWebhookTemplate(zaptest.NewLogger(t), ts.URL, tmpl, nil, time.Second)
+	require.NoError(t, err)
 
 	err = template.Execute(context.TODO(), audit.Event{
 		Type:   string(flipt.SubjectFlag),
@@ -75,13 +66,11 @@ func TestExecuter_Execute(t *testing.T) {
 }
 
 func TestExecuter_Execute_toJson_valid_Json(t *testing.T) {
-	tmpl, err := template.New("").Funcs(funcMap).Parse(`{
+	tmpl := `{
 		"type": "{{ .Type }}",
 		"action": "{{ .Action }}",
 		"payload": {{ toJson .Payload }}
-	}`)
-
-	require.NoError(t, err)
+	}`
 
 	ts := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 		w.WriteHeader(200)
@@ -89,11 +78,8 @@ func TestExecuter_Execute_toJson_valid_Json(t *testing.T) {
 
 	t.Cleanup(ts.Close)
 
-	template := &webhookTemplate{
-		url:          ts.URL,
-		bodyTemplate: tmpl,
-		httpClient:   retryablehttp.NewClient(),
-	}
+	template, err := NewWebhookTemplate(zaptest.NewLogger(t), ts.URL, tmpl, nil, time.Second)
+	require.NoError(t, err)
 
 	err = template.Execute(context.TODO(), audit.Event{
 		Type:   string(flipt.SubjectFlag),

--- a/internal/server/audit/template/leveled_logger.go
+++ b/internal/server/audit/template/leveled_logger.go
@@ -1,0 +1,47 @@
+package template
+
+import (
+	"github.com/hashicorp/go-retryablehttp"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+func NewLeveledLogger(logger *zap.Logger) retryablehttp.LeveledLogger {
+	return &LeveledLogger{logger}
+}
+
+type LeveledLogger struct {
+	logger *zap.Logger
+}
+
+func (l *LeveledLogger) Error(msg string, keyvals ...interface{}) {
+	if l.logger.Core().Enabled(zapcore.ErrorLevel) {
+		l.logger.Error(msg, l.fields(keyvals)...)
+	}
+}
+
+func (l *LeveledLogger) Info(msg string, keyvals ...interface{}) {
+	if l.logger.Core().Enabled(zapcore.InfoLevel) {
+		l.logger.Info(msg, l.fields(keyvals)...)
+	}
+}
+func (l *LeveledLogger) Debug(msg string, keyvals ...interface{}) {
+	if l.logger.Core().Enabled(zapcore.DebugLevel) {
+		l.logger.Debug(msg, l.fields(keyvals)...)
+	}
+}
+
+func (l *LeveledLogger) Warn(msg string, keyvals ...any) {
+	if l.logger.Core().Enabled(zapcore.WarnLevel) {
+		l.logger.Warn(msg, l.fields(keyvals)...)
+	}
+}
+
+func (l *LeveledLogger) fields(keyvals []any) []zap.Field {
+	fields := make([]zap.Field, 0, len(keyvals)/2)
+	for i := 0; i < len(keyvals); i += 2 {
+		k, v := keyvals[i], keyvals[i+1]
+		fields = append(fields, zap.Any(k.(string), v))
+	}
+	return fields
+}

--- a/internal/server/audit/template/leveled_logger_test.go
+++ b/internal/server/audit/template/leveled_logger_test.go
@@ -1,0 +1,34 @@
+package template
+
+import (
+	"bufio"
+	"bytes"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+func TestLeveledLogger(t *testing.T) {
+	var buffer bytes.Buffer
+	enc := zap.NewDevelopmentEncoderConfig()
+	enc.TimeKey = ""
+	encoder := zapcore.NewConsoleEncoder(enc)
+	writer := bufio.NewWriter(&buffer)
+	lzap := zap.New(zapcore.NewCore(encoder, zapcore.AddSync(writer), zapcore.DebugLevel))
+	l := NewLeveledLogger(lzap)
+	l.Debug("debug", "key", false)
+	l.Info("info", "foo", "bar")
+	l.Warn("warn", "alarm", 3.2)
+	l.Error("error", "level", 10)
+
+	require.NoError(t, lzap.Sync())
+	require.NoError(t, writer.Flush())
+	assert.Equal(t, `DEBUG	debug	{"key": false}
+INFO	info	{"foo": "bar"}
+WARN	warn	{"alarm": 3.2}
+ERROR	error	{"level": 10}
+`, buffer.String())
+}

--- a/internal/server/audit/webhook/client.go
+++ b/internal/server/audit/webhook/client.go
@@ -8,9 +8,11 @@ import (
 	"encoding/json"
 	"fmt"
 	"net/http"
+	"time"
 
 	"github.com/hashicorp/go-retryablehttp"
 	"go.flipt.io/flipt/internal/server/audit"
+	"go.flipt.io/flipt/internal/server/audit/template"
 	"go.uber.org/zap"
 )
 
@@ -40,7 +42,9 @@ func (w *webhookClient) signPayload(payload []byte) []byte {
 }
 
 // NewHTTPClient is the constructor for a HTTPClient.
-func NewWebhookClient(logger *zap.Logger, url, signingSecret string, httpClient *retryablehttp.Client) Client {
+func NewWebhookClient(logger *zap.Logger, url, signingSecret string, maxBackoffDuration time.Duration) Client {
+	httpClient := retryablehttp.NewClient()
+	httpClient.Logger = template.NewLeveledLogger(logger)
 	return &webhookClient{
 		logger:        logger,
 		url:           url,

--- a/internal/server/audit/webhook/client.go
+++ b/internal/server/audit/webhook/client.go
@@ -45,6 +45,7 @@ func (w *webhookClient) signPayload(payload []byte) []byte {
 func NewWebhookClient(logger *zap.Logger, url, signingSecret string, maxBackoffDuration time.Duration) Client {
 	httpClient := retryablehttp.NewClient()
 	httpClient.Logger = template.NewLeveledLogger(logger)
+	httpClient.RetryWaitMax = maxBackoffDuration
 	return &webhookClient{
 		logger:        logger,
 		url:           url,

--- a/internal/server/audit/webhook/client_test.go
+++ b/internal/server/audit/webhook/client_test.go
@@ -8,6 +8,7 @@ import (
 	"testing"
 	"time"
 
+	"github.com/hashicorp/go-retryablehttp"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"go.flipt.io/flipt/internal/server/audit"
@@ -16,7 +17,7 @@ import (
 )
 
 func TestConstructorWebhookClient(t *testing.T) {
-	client := NewWebhookClient(zap.NewNop(), "https://flipt-webhook.io/webhook", "", time.Second)
+	client := NewWebhookClient(zap.NewNop(), "https://flipt-webhook.io/webhook", "", 8*time.Second)
 
 	require.NotNil(t, client)
 
@@ -25,6 +26,9 @@ func TestConstructorWebhookClient(t *testing.T) {
 
 	assert.Equal(t, "https://flipt-webhook.io/webhook", whClient.url)
 	assert.Empty(t, whClient.signingSecret)
+	assert.Equal(t, 8*time.Second, whClient.httpClient.RetryWaitMax)
+	_, ok = whClient.httpClient.Logger.(retryablehttp.LeveledLogger)
+	assert.True(t, ok)
 }
 
 func TestWebhookClient(t *testing.T) {

--- a/internal/server/audit/webhook/client_test.go
+++ b/internal/server/audit/webhook/client_test.go
@@ -6,8 +6,8 @@ import (
 	"net/http"
 	"net/http/httptest"
 	"testing"
+	"time"
 
-	"github.com/hashicorp/go-retryablehttp"
 	"github.com/stretchr/testify/assert"
 	"github.com/stretchr/testify/require"
 	"go.flipt.io/flipt/internal/server/audit"
@@ -16,7 +16,7 @@ import (
 )
 
 func TestConstructorWebhookClient(t *testing.T) {
-	client := NewWebhookClient(zap.NewNop(), "https://flipt-webhook.io/webhook", "", retryablehttp.NewClient())
+	client := NewWebhookClient(zap.NewNop(), "https://flipt-webhook.io/webhook", "", time.Second)
 
 	require.NotNil(t, client)
 
@@ -39,12 +39,7 @@ func TestWebhookClient(t *testing.T) {
 
 	t.Cleanup(ts.Close)
 
-	client := &webhookClient{
-		logger:        zap.NewNop(),
-		url:           ts.URL,
-		signingSecret: "s3crET",
-		httpClient:    retryablehttp.NewClient(),
-	}
+	client := NewWebhookClient(zap.NewNop(), ts.URL, "s3crET", time.Second)
 
 	resp, err := client.SendAudit(context.TODO(), audit.Event{
 		Type:   string(flipt.SubjectFlag),

